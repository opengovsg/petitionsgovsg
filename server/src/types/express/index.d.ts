declare namespace Express {
  interface User {
    id: string
    type: string
    fullname: string
  }
}
