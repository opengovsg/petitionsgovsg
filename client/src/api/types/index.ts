export * from './post'
export * from './signature'
export * from './addressee'
export * from './auth'
