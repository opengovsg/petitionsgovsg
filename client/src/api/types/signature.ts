import { BaseModelParams, MessageResponse } from './common'

export type BaseSignatureDto = BaseModelParams & {
  comment: string | null
  userId: number
  postId: number
  fullname: string | null
}

export type CreateSignatureReqDto = { text: string }

// Returns signature ID
export type CreateSignatureResDto = MessageResponse & { data: number }
