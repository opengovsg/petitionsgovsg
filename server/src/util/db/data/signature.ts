export const mockSignature = {
  id: 4,
  hashedUserSgid:
    '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
  fullname: null,
  comment: 'I want to support this cause.',
  postId: 1,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}

export const mockSignatures = [
  {
    id: 1,
    hashedUserSgid: '1234',
    fullname: null,
    comment: null,
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 2,
    hashedUserSgid: '5678',
    fullname: 'John Doe',
    comment: null,
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 3,
    hashedUserSgid: '8765',
    fullname: null,
    comment: 'I support this cause!',
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
