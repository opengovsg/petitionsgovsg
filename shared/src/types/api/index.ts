export * from './auth'
export * from './common'
export * from './env'
export * from './post'
export * from './signature'
export * from './subscription'
