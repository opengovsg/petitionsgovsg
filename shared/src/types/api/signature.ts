export type CreateSignatureReqDto = {
  comment: string | null
  useName: boolean
  type?: string
}

// Returns signature ID
export type CreateSignatureResDto = number
