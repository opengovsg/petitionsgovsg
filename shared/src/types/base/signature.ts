import { z } from 'zod'
import { BaseModel } from './common'

export const Signature = BaseModel.extend({
  comment: z.string().nullable(),
  hashedUserSgid: z.string(),
  postId: z.string(),
  fullname: z.string().nullable(),
})

export type Signature = z.infer<typeof Signature>
