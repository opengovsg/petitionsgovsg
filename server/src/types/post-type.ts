import { Post } from '~shared/types/base'

export type PostEditType = {
  id: number
  userid: number
  summary: string
  title: string
}

export type UpdatePostRequestDto = Pick<Post, 'title' | 'summary'>
