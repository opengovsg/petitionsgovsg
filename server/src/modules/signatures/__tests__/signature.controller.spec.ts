import express from 'express'
import { StatusCodes } from 'http-status-codes'
import { ControllerHandler } from '../../../types/response-handler'
import supertest from 'supertest'
import { SignatureController } from '../signature.controller'
import { UserAuthType } from '~shared/types/api'

describe('SignatureController', () => {
  const signatureService = {
    listSignatures: jest.fn(),
    createSignature: jest.fn(),
    deleteSignature: jest.fn(),
  }
  const authService = {
    hasPermissionToEditPost: jest.fn(),
  }

  const signatureController = new SignatureController({
    signatureService,
    authService,
  })

  // Set up auth middleware to inject user
  const goodUser = { id: 1, type: UserAuthType.Public }
  let user: Express.User | undefined = goodUser
  const middleware: ControllerHandler = (req, _res, next) => {
    req.user = user
    next()
  }

  const noErrors: { errors: () => { msg: string }[] }[] = []
  let errors: { errors: () => { msg: string }[] }[] = noErrors
  const invalidateIfHasErrors: ControllerHandler = (req, _res, next) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    req['express-validator#contexts'] = errors
    next()
  }

  const postId = 1
  const postPath = `/${postId}`

  const answer = { id: 4, userId: user.id, postId, comment: 'Signature Body' }
  const answerPath = `/${answer.id}`

  beforeEach(() => {
    user = goodUser
    jest.resetAllMocks()
  })

  describe('listSignatures', () => {
    const app = express()
    app.get('/:id', signatureController.listSignatures)

    const request = supertest(app)

    it('returns OK on valid query', async () => {
      const signatures = [answer]
      signatureService.listSignatures.mockResolvedValueOnce(signatures)
      const response = await request.get(postPath)

      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(signatures)
      expect(signatureService.listSignatures).toHaveBeenCalledWith(postId)
    })

    it('returns INTERNAL_SERVER_ERROR on bad service', async () => {
      signatureService.listSignatures.mockRejectedValue(new Error())

      const response = await request.get(postPath)

      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({ message: 'Server Error' })
      expect(signatureService.listSignatures).toHaveBeenCalledWith(postId)
    })
  })

  describe('createSignature', () => {
    const app = express()
    app.use(express.json())
    app.use(middleware)
    app.use(invalidateIfHasErrors)
    app.post('/:id', signatureController.createSignature)
    const request = supertest(app)

    beforeEach(() => {
      user = goodUser
      errors = noErrors
      authService.hasPermissionToEditPost.mockResolvedValue(true)
      signatureService.createSignature.mockResolvedValue(answer.id)
    })

    it('returns BAD_REQUEST on bad request', async () => {
      errors = [
        {
          errors: () => [
            {
              msg: 'Validation Error',
            },
          ],
        },
      ]

      const response = await request
        .post(postPath)
        .send({ text: answer.comment })

      expect(response.status).toEqual(StatusCodes.BAD_REQUEST)
      expect(signatureService.createSignature).not.toHaveBeenCalled()
    })

    it('returns UNAUTHORIZED on no user', async () => {
      user = undefined

      const response = await request
        .post(postPath)
        .send({ text: answer.comment })

      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toStrictEqual({ message: 'User not signed in' })
      expect(signatureService.createSignature).not.toHaveBeenCalled()
    })

    it('returns OK on valid submission', async () => {
      const response = await request
        .post(postPath)
        .send({ text: answer.comment })

      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(answer.id)

      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { id, ...answerAttributes } = answer
      expect(signatureService.createSignature).toHaveBeenCalledWith(
        answerAttributes,
      )
    })

    it('returns INTERNAL_SERVER_ERROR on bad service', async () => {
      signatureService.createSignature.mockRejectedValue(new Error())

      const response = await request
        .post(postPath)
        .send({ text: answer.comment })

      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({ message: 'Server Error' })

      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { id, ...answerAttributes } = answer
      expect(signatureService.createSignature).toHaveBeenCalledWith(
        answerAttributes,
      )
    })
  })

  describe('deleteSignature', () => {
    const app = express()
    app.delete('/:id', signatureController.deleteSignature)
    const request = supertest(app)

    it('returns OK on valid query', async () => {
      const response = await request.delete(answerPath)

      expect(response.status).toEqual(StatusCodes.OK)
      expect(signatureService.deleteSignature).toHaveBeenCalledWith(answer.id)
    })

    it('returns INTERNAL_SERVER_ERROR on bad service', async () => {
      signatureService.deleteSignature.mockRejectedValue(new Error())

      const response = await request.delete(answerPath)

      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({ message: 'Server Error' })
      expect(signatureService.deleteSignature).toHaveBeenCalledWith(answer.id)
    })
  })
})
