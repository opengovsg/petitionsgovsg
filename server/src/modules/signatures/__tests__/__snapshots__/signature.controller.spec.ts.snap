// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignatureController checkUserHasSigned returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`SignatureController checkUserHasSigned returns OK with null body on user has not signed 1`] = `null`;

exports[`SignatureController checkUserHasSigned returns OK with user signature on user has signed 1`] = `
Object {
  "comment": "I want to support this cause.",
  "createdAt": "2022-01-01T00:00:00+00:00",
  "fullname": null,
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "id": 4,
  "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
  "type": "support",
  "updatedAt": "2022-01-01T00:00:00+00:00",
}
`;

exports[`SignatureController checkUserHasSigned returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;

exports[`SignatureController createSignature returns BAD_REQUEST on invalidated request 1`] = `Object {}`;

exports[`SignatureController createSignature returns BAD_REQUEST on user has alaready signed 1`] = `
Object {
  "message": "User has already signed this petition",
}
`;

exports[`SignatureController createSignature returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`SignatureController createSignature returns OK on valid submission with post published when signature count hits threshold 1`] = `4`;

exports[`SignatureController createSignature returns OK on valid submission with useName set to false 1`] = `4`;

exports[`SignatureController createSignature returns OK on valid submission with useName set to true 1`] = `4`;

exports[`SignatureController createSignature returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;

exports[`SignatureController listSignatures retrieves all signatures to a post with /:id 1`] = `
Array [
  Object {
    "comment": null,
    "createdAt": "2022-01-01T00:00:00+00:00",
    "fullname": null,
    "hashedUserSgid": "1234",
    "id": 1,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
  Object {
    "comment": null,
    "createdAt": "2022-01-01T00:00:00+00:00",
    "fullname": "John Doe",
    "hashedUserSgid": "5678",
    "id": 2,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
  Object {
    "comment": "I support this cause!",
    "createdAt": "2022-01-01T00:00:00+00:00",
    "fullname": null,
    "hashedUserSgid": "8765",
    "id": 3,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
]
`;

exports[`SignatureController listSignatures returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;
